lib/livebook/evaluator.ex

```diff
@@ -178,13 +207,50 @@ defmodule Livebook.Evaluator do
   end
 
   def handle_cast({:request_completion_items, send_to, ref, hint, evaluation_ref}, state) do
-    context = Map.get_lazy(state.contexts, evaluation_ref, fn -> initial_context() end)
+    context = get_context(state, evaluation_ref)
     items = Livebook.Completion.get_completion_items(hint, context.binding, context.env)
     send(send_to, {:completion_response, ref, items})
 
     {:noreply, state}
   end
 
+  @impl true
+  def handle_call({:fetch_evaluation_context, ref, cached_id}, _from, state) do
+    context = get_context(state, ref)
+
+    reply =
+      if context.id == cached_id do
+        {:error, :not_modified}
+      else
+        {:ok, context}
+      end
+
+    {:reply, reply, state}
+  end
+
+  def handle_call({:initialize_from, source_evaluator, source_evaluation_ref}, _from, state) do
+    state =
+      case Evaluator.fetch_evaluation_context(
+             source_evaluator,
+             source_evaluation_ref,
+             cached_id: state.initial_context.id
```

Comment:
Each context now has a unique id. If we keep reevaluating the first cell in a branching section, we will transfer the context only the first time.